File: track_master_project/track_master_localization/src/kalman_filter.cpp
#include "track_master_localization/kalman_filter.hpp"
using std::placeholders::_1;


KalmanFilter::KalmanFilter(const std::string name) : Node(name)
{
    mean_ = 0.0;
    variance_ = 1000.0;
    imu_angular_z_ = 0.0;
    is_first_odom_ = true;
    last_angular_z_ = 0.0;
    motion_ = 0.0;

    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>("/diff_cont/odom",10,std::bind(KalmanFilter::odomCb, this,_1));
    imu_sub_ = create_subscription<sensor_msgs::msg::Imu>("imu",10,std::bind(KalmanFilter::imuCb, this,_1));
    odom_filtered_pub_ = create_publisher<nav_msgs::msg::Odometry>("odom_filtered",10);

}

void KalmanFilter::odomCb(const nav_msgs::msg::Odometry& msg){
    odom_filtered_ = msg;

    if (is_first_odom_){
        is_first_odom_ = false;
        mean_ = msg.twist.twist.angular.z;
        last_angular_z_ = msg.twist.twist.angular.z;

    }

    // statePrediction();
    // measurementUpdate();
}

void KalmanFilter::imuCb(const sensor_msgs::msg::Imu& msg){
    imu_angular_z_ = msg.angular_velocity.z;
}

File: src/track_master_localization/track_master_localization/kalman_filter.py
#!/usr/bin/env python3

import rclpy 
from rclpy.node import Node
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu

class KalmanFilter(Node):
    def __init__(self):
        super().__init__("kalman_filter")

        self.odom_sub =self.create_subscription(Odometry,"/diff_drive_controller/odom",self.odom_cb,10)
        self.imu_sub = self.create_subscription(Imu,"imu",self.imu_cb,10)
        self.odom_filtered_pub = self.create_publisher(Odometry,"odom_filtered",10)

        #use of a gaussian distribution with huge uncertainty
        self.mean = 0.0
        self.variance = 1.0 

        # We are going to fuse only robot angular vel
        self.imu_angular_z = 0.0 
        self.is_first_odom = True
        self.last_angular_z = 0.0 

        self.motion_variance = 0.05 #relative to odometry
        self.measurement_variance = 5.0 #relative to imu

        self.motion = 0.0 #difference betwee two consecutive angular velocities
        self.filtered_odom = Odometry()

    def imu_cb(self, msg):
        self.imu_angular_z = msg.angular_velocity.z

    def odom_cb(self, msg):
        self.filtered_odom = msg  # first estimate

        if self.is_first_odom :
            self.is_first_odom = False

            self.mean = msg.twist.twist.angular.z
            self.last_angular_z = msg.twist.twist.angular.z
            return
        
        self.motion = msg.twist.twist.angular.z - self.last_angular_z
        print(self.last_angular_z)
        self.state_prediction()

        self.measurement_update()

        self.filtered_odom.twist.twist.angular.z = self.mean
        self.odom_filtered_pub.publish(self.filtered_odom)
        self.last_angular_z = msg.twist.twist.angular.z

    def measurement_update(self):
        self.mean = (self.measurement_variance * self.mean + self.variance * self.imu_angular_z ) / (self.variance + self.measurement_variance)
        self.variance = (self.variance * self.measurement_variance) / (self.variance + self.measurement_variance)

    def state_prediction(self):
        self.mean += self.motion
        self.variance += self.motion_variance

        
if __name__ == '__main__':
    rclpy.init()
    node = KalmanFilter()
    rclpy.spin(node)
    rclpy.shutdown()